### PR TITLE
Tim i06

### PR DESCRIPTION
**Pull request by Timofey-Vorobev**
Please review fixed issue #06

### CODE REVIEW

File: php/Tests/Test.php [hunk 1]
```diff
@@ -11,4 +11,37 @@ public function testReady()
     {
         $this->assertTrue(version_compare(PHP_VERSION, "7.0.0", "ge")) ;
     }
+
```
Comment (p5ych0) [inline php/Tests/Test.php]: please create separate file for this test

File: php/Tests/Test.php [hunk 2]
```diff
@@ -11,4 +11,37 @@ public function testReady()
...
+	 */
+    public function quarterDataProvider()
+    {
+        return [
+            [90, 2017, 1],		// OK, normal year
+            [91, 2017, 2],		// OK
+            [91, 2012, 1],		// OK, leap year
+            [91, 2012, 2],		// OK
+            [92, 2017, 4],		// OK
+            // failures from here
```
Comment (p5ych0) [inline php/Tests/Test.php]: separate data provider

File: php/Tests/Test.php [hunk 3]
```diff
@@ -11,4 +11,37 @@ public function testReady()
...
+    }
+
+    /**
+     * Returns OK if expected result [exp] is returned by the method being tested
+     * @param int $exp	expected number of days
+     * @param int $year year(YYYY)
+     * @param int $month (1-12)
+     * @dataProvider quarterDataProvider
+     */
+    public function testGetDaysInQuarter($exp, $year, $month)
```
Comment (p5ych0) [inline php/Tests/Test.php]: need more test cases

File: php/Tests/Test.php [hunk 4]
```diff
@@ -11,4 +11,37 @@ public function testReady()
...
+    {
+        $this->assertEquals($exp, Utils\Dates::getDaysInQuarter($year, $month));
```
Comment (p5ych0) [inline php/Tests/Test.php]: will always be true, use correct assertion

File: php/Utils/Dates.php [hunk 5]
```diff
@@ -14,8 +14,23 @@ class Dates
     */
     public static function getDaysInQuarter(int $year, int $quarter): int
     {
+        // check if the quarter is valid
+        if ($quarter < 1) {
```
Comment (p5ych0) [inline php/Utils/Dates.php]: both checks could be done at once and raise the same exception

File: php/Utils/Dates.php [hunk 6]
```diff
@@ -14,8 +14,23 @@ class Dates
...
+            throw new \InvalidArgumentException(sprintf('Quarter value must be > 0, got %d', $quarter));
+        }
+        if ($quarter > 4) {
+            throw new \InvalidArgumentException(sprintf('Quarter value must be <= 4, got %d', $quarter));
+        }
+        // do not check year
         // @TODO: create body of this function and cover with tests
+        $monthsInQuarter = 3;
```
Comment (p5ych0) [inline php/Utils/Dates.php]: this implementation takes roughly O(3+1), I want to see O(1)
decompose the task better

File: php/Tests/Test.php (hunk 4 above)
Comment (Timofey-Vorobev) [reply php/Tests/Test.php]: No, it is not - see attached test.log
=====
[test.log](https://github.com/p5ych0/probegin-test/files/1475298/test.log)

Number of errors+failures matches the number of invalid test cases. The assert works correctly

File: php/Utils/Dates.php [hunk 7]
```diff
@@ -14,8 +14,33 @@ class Dates
...
+        $endMonth = $startMonth + $monthsInQuarter;
+        $endYear = $year;
+        if ($endMonth > 12) {
+            $endMonth -= 12;
+            $endYear++;
+        }
+        $startDate = new \DateTime(sprintf('%d-%d-01', $year, $startMonth));
+        $endDate = new \DateTime(sprintf('%d-%d-01', $endYear, $endMonth));
+        $interval = $endDate->diff($startDate);
+        return $interval->days;
```
Comment (p5ych0) [inline php/Utils/Dates.php]: still expensive computations =)

**Comment by Timofey-Vorobev** [comment]
Updated

File: php/Utils/Dates.php [hunk 8]
```diff
@@ -14,8 +19,37 @@ class Dates
...
     {
-        // @TODO: create body of this function and cover with tests
+        // check if the quarter is valid
+        if ($quarter < 1 || $quarter > 4) {
+            throw new \InvalidArgumentException(sprintf('Quarter value must be in range [1..4], got %d', $quarter));
+        }
+        // do not check year validity
+        $monthsInQuarter = 3;
+        $startMonth = $monthsInQuarter * ($quarter - 1) + 1;
+        $days = self::DAYS_IN_MONTHS[$startMonth] + self::DAYS_IN_MONTHS[$startMonth + 1] + self::DAYS_IN_MONTHS[$startMonth + 2];
```
Comment (p5ych0) [inline php/Utils/Dates.php]: you don't need months in quarter, they're always the same

File: php/Utils/Dates.php [hunk 9]
```diff
@@ -14,8 +19,37 @@ class Dates
...
+        return $days;
     }
 
+    /**
+     * Returns true if a given year is leap year
+     * @param int $year
+     * @return boolean
+     * @todo make private if external use is not needed
+     */
+    public static function isLeapYear(int $year): bool
```
Comment (p5ych0) [inline php/Utils/Dates.php]: acceptable, but could have a return in single line of code

**Comment by Timofey-Vorobev** [comment]
Updated

**Pull request closed without merging**